Add visual report

juliaapolonio opened the pull request:
<!--
# sanger-tol/variantcalling pull request

Many thanks for contributing to sanger-tol/variantcalling!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/sanger-tol/variantcalling/tree/main/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Output Documentation in `docs/output.md` is updated.


This PR is to address issue #89 (publish visual_report files). To test these changes, I ran the test profile for aligned reads and verified that the report was emitted on the outdir/variantcalling folder.
The changes I made were:
- adding an emission in the deepvariant module/subworkflow and its stub;
- adding the publishdir for deepvariant process on modules.config
- adding the report description on meta.yml

nf-core linting failed, but all messages were related to outdated schemas of the pipeline.

And answering the question on the issue, MultiQC does not support deepvariant (https://docs.seqera.io/multiqc/modules/).

juliaapolonio committed:
add visual report emmission

juliaapolonio committed:
publish the emission on the outdir

juliaapolonio committed:
add visual report as a emission of the subworkflow

juliaapolonio committed:
add output description to docs

juliaapolonio committed:
fixing target origin branch w/ change in docs

muffato commented:
Hi @juliaapolonio .

Thanks a lot for having a go at this issue ! I tested your branch and found a couple of things that may need a further change.

The `publishDir` rule you added needs a tweak. As it is, it's copying the entire output directory of `DEEPVARIANT_CALLER:DEEPVARIANT` into `variant_calling/`. That means including the VCF files. The problem is that the pipeline split the input genome into chunks to run DeepVariant in parallel, and then merges the VCF files together. With your change, the `variant_calling/` directory has got both the pre- and post-merge VCF files. You need to add a filter to only copy the html file.

Then, I believe you don't need to add an output channel for the html files. I think will change in the next version of Nextflow but currently, `publishDir` works on the task directory itself with no consideration for how the output channels are defined. Try without the extra channel and see if it still works.

Finally, coming back to the notion of genome chunks. We'll actually have 1 HTML file per chunk, which could be a lot for large genomes. What about putting the HTML files in a sub-directory e.g. `variant_calling/deepvariant_reports/` ?

And thank you for checking MultiQC out.

Best,
Matthieu

juliaapolonio committed:
remove output channels of the html report

juliaapolonio committed:
create subdir for deepvariant reports; filter publish to html files

juliaapolonio committed:
emit visual_report as output

juliaapolonio commented:
Hi @muffato

Thanks for the review! Here's what I did:

1) I put a pattern in the publishdir to filter only the visual_report files, and removed the saveas option since I was using this specific pattern. I've also added a deepvariant report subfolder

2) I tried removing the HTML output channel from the subworkflow and the module, but it didn't publish the report, so I put it back in the module. 
I searched in the Nextflow documentation but couldn't find anything about the need to output what you publish. Do you know if I missed something? My Nextflow version is 24.10.5

Best,
Julia

muffato reviewed (commented):
> 1. I put a pattern in the publishdir to filter only the visual_report files, and removed the saveas option since I was using this specific pattern. I've also added a deepvariant report subfolder

That's an elegant solution. We should be using `pattern` more often !

> 2. I tried removing the HTML output channel from the subworkflow and the module, but it didn't publish the report, so I put it back in the module.
>    I searched in the Nextflow documentation but couldn't find anything about the need to output what you publish. Do you know if I missed something? My Nextflow version is 24.10.5

Ah, then I was wrong ! I just checked and also found the files would not be published without the output channel.

Since you modified `modules/nf-core/deepvariant/rundeepvariant/main.nf` which is controlled by nf-core, you need to run the `nf-core modules patch` command to update a `.diff` file in the same directory. Otherwise nf-core complains that the file has got local modifications.

juliaapolonio committed:
add diff file for deepvariant module

juliaapolonio committed:
change tabulations to whitespace

juliaapolonio committed:
place the deepvariant output description correctly in the docs

muffato reviewed (approved):
Super, all good !
Well done @juliaapolonio
